### PR TITLE
reuse an existing partition if possible in apply_system

### PR DESCRIPTION
**Pull request by mwhudson**
This required a bit of a refactor which lead to rewriting tests as well as adding a new one.

### CODE REVIEW

**Comment by mwhudson** [comment]
grumble something i have done has made the integration tests flaky :(

File: subiquity/server/controllers/filesystem.py [hunk 1]
```diff
@@ -452,43 +452,60 @@ def _offsets_and_sizes_for_system(self):
...
-            self._add_structure(
-                disk=disk, offset=offset, size=size, structure=structure,
-                is_last=structure == volume.structure[-1])
+            if (offset, size) in parts_by_offset_size:
+                part = parts_by_offset_size[(offset, size)]
+            else:
+                if structure.role == snapdapi.Role.SYSTEM_DATA and \
+                   structure == volume.structure[-1]:
+                    gap = gaps.largest_gap(disk)
+                    size = gap.size - (offset - gap.offset)
```
Comment (dbungert) [inline subiquity/server/controllers/filesystem.py:492]: This is just trying to use the rest of the disk, right?  Why the decrease in size?  Wouldn't it just be gap.size?

File: subiquity/server/snapdapi.py [hunk 2]
```diff
@@ -142,7 +142,10 @@ class VolumeStructure:
     update: VolumeUpdate = attr.Factory(VolumeUpdate)
 
     def gpt_part_uuid(self):
-        return self.type.split(',', 1)[1].upper()
+        if ',' in self.type:
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/snapdapi.py:145]: I hope not... Maybe I should just add code that rejects a non-gpt schema.

File: subiquity/server/controllers/filesystem.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/server/controllers/filesystem.py:492]: Because the offset in the gadget isn't necessarily immediately after the preceding structure.

Comment (dbungert) [reply subiquity/server/controllers/filesystem.py:492]: I think that works then, so long as the preceding structure obeys 1MiB alignment rules.

**Review by dbungert — APPROVED** [review]

**Comment by mwhudson** [comment]
This contains another unrelated fix, but with this branch I was able to install an encrypted system! Hurrah!

**Pull request merged**